libp2p/src/lib.rs

```diff
@@ -223,7 +224,7 @@ pub async fn development_transport(
         .authenticate(noise::NoiseAuthenticated::xx(&keypair).unwrap())
         .multiplex(core::upgrade::SelectUpgrade::new(
             yamux::YamuxConfig::default(),
-            mplex::MplexConfig::default(),
+            yamux::YamuxConfig::default(),
         ))
```

Comment:
Removing this would be a breaking change, please leave as is and add `#[allow(deprecated)]` where necessary.

Reply:
@thomaseizinger As `mplex` is deprecated, this conditional compilation block put the code after that as inactive and I don't quite understand where I can put a `#[allow(deprecated)]` to solve this issue, WDT? 

```rust
#[cfg(all(
    not(target_arch = "wasm32"),
    any(
        all(feature = "tcp-async-io", feature = "dns-async-std"),
        all(feature = "tcp", feature = "dns", feature = "async-std")
    ),
    feature = "websocket",
    feature = "noise",
    feature = "mplex",
    feature = "yamux"
))]
```

Reply:
We just want to render a deprecation warning for users. All code conditional on the `mplex` feature should just stay as is. This function references `mplex`, which will trigger the deprecation warning so you will need `#[allow(deprecated)]` on top of the function. The position relative to the `cfg` attribute shouldn't matter.

Reply:
@thomaseizinger Sorry I must be missing something here, I'm not super familiar with configuration conditional checks but I'd like to understand. I expected something like that:

```rust
#[cfg(all(
    not(target_arch = "wasm32"),
    any(
        all(feature = "tcp-async-io", feature = "dns-async-std"),
        all(feature = "tcp", feature = "dns", feature = "async-std")
    ),
    feature = "websocket",
    feature = "noise",
    feature = "mplex",
    feature = "yamux"
))]
#[cfg_attr(
    all(
        any(feature = "tcp-async-io", feature = "dns-async-std"),
        not(feature = "async-std")
    ),
    deprecated(
        since = "0.49.0",
        note = "The `tcp-async-io` and `dns-async-std` features are deprecated. Use the new `tcp` and `dns` features together with the `async-std` feature."
    )
)]
pub async fn development_transport(
    keypair: identity::Keypair,
) -> std::io::Result<core::transport::Boxed<(PeerId, core::muxing::StreamMuxerBox)>> {
    let transport = {
        let dns_tcp = dns::DnsConfig::system(tcp::async_io::Transport::new(
            tcp::Config::new().nodelay(true),
        ))
        .await?;
        let ws_dns_tcp = websocket::WsConfig::new(
            dns::DnsConfig::system(tcp::async_io::Transport::new(
                tcp::Config::new().nodelay(true),
            ))
            .await?,
        );
        dns_tcp.or_transport(ws_dns_tcp)
    };

    Ok(transport
        .upgrade(core::upgrade::Version::V1)
        .authenticate(noise::NoiseAuthenticated::xx(&keypair).unwrap())
        .multiplex(core::upgrade::SelectUpgrade::new(
            yamux::YamuxConfig::default(),
            #[allow(deprecated)]
            mplex::MplexConfig::default(),
        ))
        .timeout(std::time::Duration::from_secs(20))
        .boxed())
}
```
But in this case the whole block is marked as inactive because I marked the `mplex` module as depreciated at the beginning of the file:
```rust
#[cfg(feature = "mplex")]
#[deprecated(
    note = "`mplex` is not recommended anymore. Please use `yamux` instead or depend on `libp2p-mplex` directly if you need it for legacy use cases."
)]
pub mod mplex {
    pub use libp2p_mplex::*;
}
```
However, of course if I remove the `feature = "mplex"` line of the `cfg` such as 
```rust
#[cfg(all(
    not(target_arch = "wasm32"),
    any(
        all(feature = "tcp-async-io", feature = "dns-async-std"),
        all(feature = "tcp", feature = "dns", feature = "async-std")
    ),
    feature = "websocket",
    feature = "noise",
    feature = "yamux"
))]
```
The `mplex` module is then undeclared in 
```rust
.multiplex(core::upgrade::SelectUpgrade::new(
            yamux::YamuxConfig::default(),
            #[allow(deprecated)]
            mplex::MplexConfig::default(),
        ))
```
Do you have an idea here? I suppose so, this is a stupid stuff but I was not able to figure it out :)

Reply:
> But in this case the whole block is marked as inactive because I marked the `mplex` module as depreciated at the beginning of the file:

Do you use a jetbrains IDE? Or what do you mean with inactive?

Reply:
@thomaseizinger I use VSCode. In fact, I was expected what I've just pushed as correct but the CI IPFS integration test fails due to `no development_transport in the root`. And from my understanding and tests, this is because the `development_transport` function is not compiled due to the `mplex` deprecated module even if I put `#[allow(deprecated)]` in here.

Reply:
See other comments!